File: client-sdk/python/src/synadia_ai/agents/identity/id_sig.py
"""``id_sig`` — the registration signature that makes the registered agent ID verifiable.

Spec "Registration". Signed input (never sent; note the field order —
``user`` first, unlike ``AGENT-SENDER-V1``)::

    AGENT-ID-V1\\n{user_nkey}\\n{account}\\n{agent}\\n{owner}\\n{prompt_subject}\\n

A verifier rebuilds it from the ``user_nkey``, ``account``, ``agent``,
``owner`` metadata and the ``prompt`` endpoint's subject, all from one
``$SRV.INFO`` reply. Verification is synchronous.
"""

from __future__ import annotations

from collections.abc import Mapping

from ..errors import IdentityError
from ._nkeys import base64url_decode, base64url_encode, verify_with_public_key
from .agent_id import AgentId, assert_valid_user_key
from .signer import SenderSigner, maybe_await

AGENT_ID_SIGNED_INPUT_TAG = "AGENT-ID-V1"

#: The three metadata keys the host registers.
METADATA_USER_NKEY = "user_nkey"
METADATA_ACCOUNT = "account"
METADATA_ID_SIG = "id_sig"
IDENTITY_METADATA_KEYS: frozenset[str] = frozenset(
    {METADATA_USER_NKEY, METADATA_ACCOUNT, METADATA_ID_SIG}
)

_SIGNATURE_BYTES = 64


def build_agent_id_signed_input(
    *, user: str, account: str, agent: str, owner: str, prompt_subject: str
) -> bytes:
    """The exact bytes that are signed (never sent)."""
    return (
        f"{AGENT_ID_SIGNED_INPUT_TAG}\n{user}\n{account}\n{agent}\n{owner}\n{prompt_subject}\n"
    ).encode()


async def sign_agent_id(
    *, signer: SenderSigner, id: AgentId, agent: str, owner: str, prompt_subject: str
) -> str:
    """Produce the base64url ``id_sig`` value for the registration metadata."""
    data = build_agent_id_signed_input(
        user=id.user, account=id.account, agent=agent, owner=owner, prompt_subject=prompt_subject
    )
    return base64url_encode(await maybe_await(signer.sign(data)))


def verify_agent_id(metadata: Mapping[str, str], prompt_subject: str) -> bool:
    """Verify a service record's ``id_sig`` against its own ``prompt`` endpoint subject.

    ``False`` when any field is missing, the key or signature is
    malformed, or the signature does not verify.
    """
    user = metadata.get(METADATA_USER_NKEY)
    account = metadata.get(METADATA_ACCOUNT)
    agent = metadata.get("agent")
    owner = metadata.get("owner")
    id_sig = metadata.get(METADATA_ID_SIG)
    if user is None or account is None or agent is None or owner is None or id_sig is None:
        return False
    try:
        assert_valid_user_key(user)
        sig = base64url_decode(id_sig)
    except (IdentityError, ValueError):
        return False
    if len(sig) != _SIGNATURE_BYTES:
        return False
    data = build_agent_id_signed_input(
        user=user, account=account, agent=agent, owner=owner, prompt_subject=prompt_subject
    )
    return verify_with_public_key(user, data, sig)


__all__ = [
    "AGENT_ID_SIGNED_INPUT_TAG",
    "IDENTITY_METADATA_KEYS",
    "METADATA_ACCOUNT",
    "METADATA_ID_SIG",
    "METADATA_USER_NKEY",
    "build_agent_id_signed_input",
    "sign_agent_id",
    "verify_agent_id",
]

File: client-sdk/typescript/src/identity/id-sig.ts
// `id_sig` — the registration signature that makes the agent ID in the
// service metadata verifiable (spec "Registration"). Signed input (never
// sent; note the field order — `user` first, unlike `AGENT-SENDER-V1`):
//
//   AGENT-ID-V1\n{user_nkey}\n{account}\n{agent}\n{owner}\n{prompt_subject}\n
//
// A verifier rebuilds it from the `user_nkey`, `account`, `agent`, `owner`
// metadata and the `prompt` endpoint's subject, all from one `$SRV.INFO`
// reply. Verification is synchronous (nkeys `verify` is sync).

import { agentIdAccount, agentIdUser, assertValidUserKey, type AgentId } from "./agent-id.js";
import { base64UrlDecode, base64UrlEncode, utf8, verifyWithPublicKey } from "./crypto.js";
import type { SenderSigner } from "./signer.js";

export const AGENT_ID_SIGNED_INPUT_TAG = "AGENT-ID-V1";

/** The three metadata keys the host registers. */
export const IDENTITY_METADATA_KEYS = Object.freeze({
  userNkey: "user_nkey",
  account: "account",
  idSig: "id_sig",
});

export interface AgentIdSignedInputFields {
  readonly user: string;
  readonly account: string;
  readonly agent: string;
  readonly owner: string;
  readonly promptSubject: string;
}

export function buildAgentIdSignedInput(f: AgentIdSignedInputFields): Uint8Array {
  return utf8.encode(
    `${AGENT_ID_SIGNED_INPUT_TAG}\n${f.user}\n${f.account}\n${f.agent}\n${f.owner}\n${f.promptSubject}\n`,
  );
}

export interface SignAgentIdOptions {
  readonly signer: SenderSigner;
  readonly id: AgentId;
  readonly agent: string;
  readonly owner: string;
  readonly promptSubject: string;
}

/** Produce the base64url `id_sig` value for the registration metadata. */
export async function signAgentId(opts: SignAgentIdOptions): Promise<string> {
  const input = buildAgentIdSignedInput({
    user: agentIdUser(opts.id),
    account: agentIdAccount(opts.id),
    agent: opts.agent,
    owner: opts.owner,
    promptSubject: opts.promptSubject,
  });
  return base64UrlEncode(await opts.signer.sign(input));
}

/**
 * Verify a service record's `id_sig` against its own `prompt` endpoint
 * subject. `false` when any field is missing, the key or signature is
 * malformed, or the signature does not verify.
 */
export function verifyAgentId(
  metadata: Readonly<Record<string, string | undefined>>,
  promptSubject: string,
): boolean {
  const user = metadata[IDENTITY_METADATA_KEYS.userNkey];
  const account = metadata[IDENTITY_METADATA_KEYS.account];
  const agent = metadata["agent"];
  const owner = metadata["owner"];
  const idSig = metadata[IDENTITY_METADATA_KEYS.idSig];
  if (
    user === undefined ||
    account === undefined ||
    agent === undefined ||
    owner === undefined ||
    idSig === undefined
  ) {
    return false;
  }
  let sig: Uint8Array;
  try {
    assertValidUserKey(user);
    sig = base64UrlDecode(idSig);
  } catch {
    return false;
  }
  if (sig.length !== 64) return false;
  return verifyWithPublicKey(
    user,
    buildAgentIdSignedInput({ user, account, agent, owner, promptSubject }),
    sig,
  );
}
